File: ios/localasset-core/Sources/LocalAssetCore/internal/AdapterRegistry.swift
struct AdapterRegistry {
    private let adapters: [SchemeAdapter]

    init(adapters: [SchemeAdapter]) {
        self.adapters = adapters
    }

    func select(url: String) throws -> SchemeAdapter {
        guard let adapter = adapters.first(where: { $0.canHandle(url: url) }) else {
            throw ResourceException(.parseError, "no adapter can handle url")
        }
        return adapter
    }
}

File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/internal/AdapterRegistry.kt
package io.github.xesam.android.localasset.core.internal

import io.github.xesam.android.localasset.core.api.SchemeAdapter
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException

class AdapterRegistry(
    private val adapters: List<SchemeAdapter>,
) {
    fun select(url: String): SchemeAdapter = adapters.firstOrNull { it.canHandle(url) }
        ?: throw ResourceException(ResourceErrorCategory.PARSE_ERROR, "no adapter can handle url")
}
